tokenizers/src/models/bpe/trainer.rs

```diff
@@ -734,4 +735,41 @@ mod tests {
         .collect();
         assert_eq!(model.merges, expected_merges);
     }
+    #[test]
+    fn bpe_test_max_token_length_16() {
+        let max_token_length = 16;
+        let long_word_counts: HashMap<String, u32> = [
+            ("singlelongtokenwithoutcasechange", 2),
+            ("singleLongTokenWithCamelCaseChange", 2),
+            ("Longsingletokenwithpunctu@t!onwithin", 2),
+            ("Anotherlongsingletokenwithnumberw1th1n", 2),
+            ("짧은한글문자열짧은한", 2),             // korean 10 char
+            ("긴한글문자열긴한글문자열긴한글문", 2), // korean 16 char
+            ("短字符串短字符串短字", 2),             //simplified chinese 10 char
+            ("长字符串长字符串长字符串长字符串", 2), // simp. chinese 16 char
+            ("短い文字列短い文字列", 2),             // japanese 10 char
+            ("長い文字列長い文字列長い文字列長", 2), // japanese 16 char
+            ("so", 2),
+            ("GPT-2", 2),
+        ]
+        .iter()
+        .map(|(key, value)| (key.to_string(), *value))
+        .collect();
+        let trainer = BpeTrainer::builder()
+            .max_token_length(max_token_length)
+            .show_progress(false)
+            .min_frequency(0)
+            .build();
+        let mut model = BPE::default();
+        trainer.do_train(&long_word_counts, &mut model).unwrap();
+        let vocab = model.get_vocab();
+        for token in vocab.keys() {
+            assert!(
+                token.chars().count() <= max_token_length,
+                "token too long : {} , chars().count() = {}",
+                token,
+                token.chars().count()
```

Comment:
Thats a great test ! Good idea on the UTF-8 intense examples ! 

Could you add the explicit vocab as a result ? It makes the test more readable and more robust.
Let's keep the functional part too, but having the explicit values prevents "under the radar" bugs, where behavior is modified unintentionally.

Reply:
What do you mean by keep explicit vocabulary?
Do you mean the tokens learned by the model under that setting?

In that case I propose we make separate tests.

The current test could serve as the basis for a robust test that checks across multiple max_token_length values and added vocabulary.
I wanted to add Arabic but it renders differently depending on the IDE (due to  left to right script) but exactly because of such behavior it would be useful to have in the future.

Another test could be added that takes a look into individual vocabularies learned, but this would be more rigid since adding more words to be learned would also change the vocabs, so we'd need to have new known good tokens to compare with.

Reply:
No 1 test, but explicit assert.

`assert_eq!(tokenizer.get_vocab(), HashMapFrom([....]))`

It's really much better imo. Tests don't change that often, and I have seen many bugs be silent for too long for this lack of explicit value testing.

Reply:
Hmm so you feel like the test already has enough unicode coverage?


What about,

Leave current test (#1)

 change current test with explicit assert compare (#2)

This way we could add any new unicode we want on #1 and #2 can serve as explicit coverage.

Is this too many tests for this feature?
Considering the breadth of this feature i think this warrants at least two.(hopefully three) But if you want it to be one, I'll try to pack as much into it